python/tvm/auto_scheduler/search_policy.py

```diff
@@ -61,6 +62,71 @@ def __init__(self, filename):
         self.__init_handle_by_constructor__(_ffi_api.PreloadMeasuredStates, filename)
 
 
+@tvm._ffi.register_object("auto_scheduler.PreloadCustomSketchRule")
+class PreloadCustomSketchRule(SearchCallback):
+    """
+    A SearchCallback for SketchSearchPolicy that allowing users to add
+    custom sketch rule.
+
+    Notes
+    -----
+    This is an advanced feature. Make sure you're clear how it
+    works and this should only be used in SketchSearchPolicy.
+
+    Parameters
+    ----------
+    meet_condition_func: Function
+        A function with `(policy, state, stage_id) -> int`
+    apply_func: Function
+        A function with `(policy, state, stage_id) -> [[State, int], ...]`
+    """
+
+    CONDITION_NUM = {"pass": 0, "apply": 1, "apply_and_skip_rest": 2}
+
+    def __init__(self, meet_condition_func, apply_func, rule_name="CustomSketchRule"):
+        self.__init_handle_by_constructor__(
+            _ffi_api.PreloadCustomSketchRule, meet_condition_func, apply_func, rule_name
+        )
+
+
+CUSTOM_SKETCH_REGISTRY = {}
+
+
+def register_custom_sketch_func(compute_name, func=None):
+    """Helper decorator to register custom sketch functions easily."""
+    global CUSTOM_SKETCH_REGISTRY
+
+    if callable(compute_name):
+        func = compute_name
+        compute_name = func.__name__
+
+    if not isinstance(compute_name, str):
+        raise ValueError("expect string function name")
+
+    def register(myf):
+        if compute_name in CUSTOM_SKETCH_REGISTRY:
+            raise RuntimeError(
+                "Custom Sketch for %s has been registered for compute already" % compute_name
+            )
+
+        def meet_condition_func(policy, state, stage_id):
+            state = State(state, policy.search_task.compute_dag)
+            if state.stages[stage_id].op.name == compute_name:
+                return PreloadCustomSketchRule.CONDITION_NUM["apply_and_skip_rest"]
+            return PreloadCustomSketchRule.CONDITION_NUM["pass"]
```

Comment:
Does this mean if users use this decorator to register the custom rule, they cannot customize the condition function?

Reply:
Actually I'm not sure if this decorator is really needed.
Full custom sketch support(user defined condtion function and apply function) can be acquired by creating PreloadCustomSketchRule object directly, and the decorator is used as a fast path.

Reply:
The fast path seems not that useful to me. In this case I think we could just remove the decorator so that the implementation could be simplified a lot.